File: mac-app/QuartzTeachers/Catalogs/CourseNameCatalog.swift
import Foundation

/// Looks up Ontario course names by course code, using the same
/// `ontario_secondary_courses.json` the command-line wizard consults
/// (bundled as an app resource). The wizard offers a formal name and a
/// short name for known codes; the app does the same.
enum CourseNameCatalog {

    // MARK: - Stored properties

    static let entries: [String: CourseNames] = loadEntries()

    // MARK: - Functions

    /// The known names for a course code, or nil for unknown codes
    /// (including club codes). Lookup is case-insensitive, matching the
    /// wizard's behaviour of uppercasing the code first.
    static func names(forCode code: String) -> CourseNames? {
        let normalized: String = code.trimmingCharacters(in: .whitespaces).uppercased()
        return entries[normalized]
    }

    /// The name a new course starts with when its code is recognised, or
    /// nil for a code the lookup does not know.
    ///
    /// It is the SHORT name. The formal name is the ministry's string —
    /// "Introduction to Computer Science, Grade 11, U" — which is correct
    /// and which almost nobody wants on a website: it is what a teacher
    /// then has to delete before typing what they actually call the
    /// course. The short name is closer to that in nearly every case, and
    /// the formal one is still a single button away for the teacher who
    /// wants it. Defaulting to the longer string only to have it thrown
    /// away is the wrong way round.
    static func defaultName(forCode code: String) -> String? {
        return names(forCode: code)?.short
    }

    private static func loadEntries() -> [String: CourseNames] {
        var result: [String: CourseNames] = [:]
        let catalogFiles: [String] = [
            "ontario_secondary_courses",
            "british_columbia_secondary_courses"
        ]
        for fileName in catalogFiles {
            guard let url = Bundle.main.url(forResource: fileName, withExtension: "json") else {
                continue
            }
            guard let data = try? Data(contentsOf: url) else {
                continue
            }
            guard let decoded = try? JSONSerialization.jsonObject(with: data) else {
                continue
            }
            guard let dictionary = decoded as? [String: Any] else {
                continue
            }

            for (code, rawEntry) in dictionary {
                guard let entry = rawEntry as? [String: Any] else {
                    continue
                }
                guard let formalName = entry["formal_name"] as? String else {
                    continue
                }
                guard let shortName = entry["short_name"] as? String else {
                    continue
                }
                result[code] = CourseNames(formal: formalName, short: shortName)
            }
        }
        return result
    }
}

/// The two name variants the lookup file offers for each course code.
struct CourseNames: Equatable {

    // MARK: - Stored properties

    let formal: String
    let short: String
}

File: windows-app/Plantoir.Core/Catalogs/CourseNameCatalog.cs
using System.Text.Json.Nodes;

namespace Plantoir.Core.Catalogs;

public sealed record CourseNames(string Formal, string Short);

/// <summary>
/// Ontario course-name lookup from support/ontario_secondary_courses.json —
/// the same file the CLI wizard consults. Codes are trimmed and uppercased
/// before lookup; unknown and club codes return null.
/// </summary>
public sealed class CourseNameCatalog
{
    private readonly Dictionary<string, CourseNames> _entries = new();

    public static CourseNameCatalog Load(params string[] jsonPaths)
    {
        var catalog = new CourseNameCatalog();
        foreach (var jsonPath in jsonPaths)
        {
            if (!File.Exists(jsonPath)) continue;
            try
            {
                if (JsonNode.Parse(File.ReadAllBytes(jsonPath)) is JsonObject root)
                    foreach (var (code, value) in root)
                    {
                        if (value is not JsonObject entry) continue;
                        if (entry["formal_name"] is not JsonValue f || !f.TryGetValue<string>(out string? formal)) continue;
                        if (entry["short_name"] is not JsonValue s || !s.TryGetValue<string>(out string? shortName)) continue;
                        catalog._entries[code] = new CourseNames(formal, shortName);
                    }
            }
            catch { }
        }
        return catalog;
    }

    public CourseNames? Names(string code)
    {
        string key = code.Trim().ToUpperInvariant();
        return _entries.TryGetValue(key, out CourseNames? names) ? names : null;
    }

    public string? DefaultName(string code) => Names(code)?.Short;

    public int Count => _entries.Count;

    /// <summary>Every known code and its names, for a picker to search over.</summary>
    public IEnumerable<(string Code, CourseNames Names)> AllEntries() =>
        _entries.Select(e => (e.Key, e.Value));
}
